File: runtime/src/agents_runtime/repository/moments.py
"""A verdade comercial com validade (§3.3.4 `commercial_moments`).

"Ativo" é computado AQUI, a cada leitura, contra o relógio do banco — nunca
gravado (invariante §3.4-2). Um momento aprovado cuja janela passou simplesmente
não volta desta query, e um kill switch (`killed_at`) vence a janela mesmo com
`ends_at` no futuro.

Escopo por RLS: a conexão do worker está trancada na org do turno, então não
há `where organization_id` aqui — a policy é o filtro (CLAUDE.md).
"""

import psycopg

from agents_runtime.commerce.moments import Moment


async def load_active_moments(conn: psycopg.AsyncConnection) -> tuple[Moment, ...]:
    """Os momentos ativos AGORA, maior prioridade primeiro.

    A ordem importa: sobrepostos somam fatos, mas a postura promocional
    (claim/offer) é a do PRIMEIRO desta lista (§3.3.4).
    """
    cursor = await conn.execute(
        """
        select id, name, public_claim, offer, temp_facts, forbidden,
               priority, template_readiness
          from public.commercial_moments
         where status = 'approved'
           and killed_at is null
           and now() between starts_at and ends_at
         order by priority desc, starts_at
        """
    )
    rows = await cursor.fetchall()
    return tuple(
        Moment(
            id=row[0],
            name=row[1],
            public_claim=row[2],
            offer=row[3] or {},
            temp_facts=tuple(str(fact) for fact in (row[4] or ())),
            forbidden=tuple(row[5] or ()),
            priority=row[6],
            template_readiness=row[7] or {},
        )
        for row in rows
    )

File: src/lib/ai/moments.ts
// Momentos comerciais (core/agentes-por-evento.md §1.5/§3.3.4) — o que a UI
// e as rotas compartilham. "Ativo" NUNCA é gravado: é computado por relógio
// (approved + agora dentro da janela + não morto) — inclusive aqui na UI.

export interface CommercialMoment {
  id: string
  organization_id: string
  store_id: string | null
  name: string
  starts_at: string
  ends_at: string
  public_claim: string | null
  offer: { kind: 'none' | 'percent' | 'fixed' | 'free_shipping'; value?: number; coupon_code?: string; auto_apply?: boolean }
  exclusions: Record<string, unknown>
  temp_facts: string[]
  forbidden: string[]
  priority: number
  template_readiness: Record<string, { template_name?: string; language?: string; status?: string }>
  status: 'draft' | 'approved' | 'killed'
  killed_at: string | null
  created_at: string
  updated_at: string
}

const EDITABLE_FIELDS = [
  'name',
  'starts_at',
  'ends_at',
  'public_claim',
  'offer',
  'exclusions',
  'temp_facts',
  'forbidden',
  'priority',
  'template_readiness',
] as const

export function pickEditableMomentFields(body: Record<string, any>) {
  const row: Record<string, any> = {}
  for (const field of EDITABLE_FIELDS) {
    if (body[field] !== undefined) row[field] = body[field]
  }
  return row
}

export type ComputedState = 'vigente' | 'agendado' | 'encerrado' | 'morto' | 'rascunho'

export function computedState(moment: CommercialMoment, now = new Date()): ComputedState {
  if (moment.killed_at) return 'morto'
  if (moment.status !== 'approved') return 'rascunho'
  if (now < new Date(moment.starts_at)) return 'agendado'
  if (now > new Date(moment.ends_at)) return 'encerrado'
  return 'vigente'
}
